metadata-ingestion/src/datahub/ingestion/source/feast_repository.py

```diff
@@ -0,0 +1,371 @@
+from typing import Dict, Iterable, List, Tuple, Union
+
+from feast import (
+    BigQuerySource,
+    Entity,
+    Feature,
+    FeatureStore,
+    FeatureView,
+    FileSource,
+    KafkaSource,
+    KinesisSource,
+    OnDemandFeatureView,
+    ValueType,
+)
+from feast.data_source import DataSource, RequestDataSource
+
+import datahub.emitter.mce_builder as builder
+from datahub.configuration.common import ConfigModel
+from datahub.emitter.mce_builder import DEFAULT_ENV
+from datahub.ingestion.api.common import PipelineContext
+from datahub.ingestion.api.source import Source, SourceReport
+from datahub.ingestion.api.workunit import MetadataWorkUnit
+from datahub.metadata.com.linkedin.pegasus2avro.common import MLFeatureDataType
+from datahub.metadata.com.linkedin.pegasus2avro.metadata.snapshot import (
+    MLFeatureSnapshot,
+    MLFeatureTableSnapshot,
+    MLPrimaryKeySnapshot,
+)
+from datahub.metadata.com.linkedin.pegasus2avro.mxe import MetadataChangeEvent
+from datahub.metadata.schema_classes import (
+    BrowsePathsClass,
+    MLFeaturePropertiesClass,
+    MLFeatureTablePropertiesClass,
+    MLPrimaryKeyPropertiesClass,
+)
+
+_field_type_mapping: Dict[ValueType, str] = {
+    ValueType.UNKNOWN: MLFeatureDataType.UNKNOWN,
+    ValueType.BYTES: MLFeatureDataType.BYTE,
+    ValueType.STRING: MLFeatureDataType.TEXT,
+    ValueType.INT32: MLFeatureDataType.ORDINAL,
+    ValueType.INT64: MLFeatureDataType.ORDINAL,
+    ValueType.DOUBLE: MLFeatureDataType.CONTINUOUS,
+    ValueType.FLOAT: MLFeatureDataType.CONTINUOUS,
+    ValueType.BOOL: MLFeatureDataType.BINARY,
+    ValueType.UNIX_TIMESTAMP: MLFeatureDataType.TIME,
+    ValueType.BYTES_LIST: MLFeatureDataType.SEQUENCE,
+    ValueType.STRING_LIST: MLFeatureDataType.SEQUENCE,
+    ValueType.INT32_LIST: MLFeatureDataType.SEQUENCE,
+    ValueType.INT64_LIST: MLFeatureDataType.SEQUENCE,
+    ValueType.DOUBLE_LIST: MLFeatureDataType.SEQUENCE,
+    ValueType.FLOAT_LIST: MLFeatureDataType.SEQUENCE,
+    ValueType.BOOL_LIST: MLFeatureDataType.SEQUENCE,
+    ValueType.UNIX_TIMESTAMP_LIST: MLFeatureDataType.SEQUENCE,
+    ValueType.NULL: MLFeatureDataType.USELESS,
```

Comment:
The USELESS type should be for unique data with no meaningful use — for instance, UUIDs. The `UNKNOWN` type would be a better fit.

Reply:
Fixed! I was reading this: `NULL` as `USELESS` value for model inputs.